src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexPrefixAnalyzer.cs

```diff
@@ -601,94 +590,116 @@ public static (RegexNode LoopNode, (char Char, string? String, char[]? Chars) Li
             return null;
         }
 
-        /// <summary>Takes a RegexTree and computes the leading anchor that it encounters.</summary>
-        public static int FindLeadingAnchor(RegexTree tree)
+        /// <summary>Computes the leading anchor of a node.</summary>
+        public static RegexNodeKind FindLeadingAnchor(RegexNode node) =>
+            FindLeadingOrTrailingAnchor(node, leading: true);
+
+        /// <summary>Computes the leading anchor of a node.</summary>
+        public static RegexNodeKind FindTrailingAnchor(RegexNode node) =>
+            FindLeadingOrTrailingAnchor(node, leading: false);
+
+        /// <summary>Computes the leading or trailing anchor of a node.</summary>
+        private static RegexNodeKind FindLeadingOrTrailingAnchor(RegexNode node, bool leading)
         {
-            RegexNode curNode = tree.Root;
-            RegexNode? concatNode = null;
-            int nextChild = 0;
+            if (!StackHelper.TryEnsureSufficientExecutionStack())
+            {
+                // We only recur for alternations, but with a really deep nesting of alternations we could potentially overflow.
+                // In such a case, simply stop searching for an anchor.
+                return RegexNodeKind.Unknown;
+            }
 
             while (true)
             {
-                switch (curNode.Kind)
+                switch (node.Kind)
                 {
                     case RegexNodeKind.Bol:
-                        return Bol;
-
                     case RegexNodeKind.Eol:
-                        return Eol;
-
-                    case RegexNodeKind.Boundary:
-                        return Boundary;
-
-                    case RegexNodeKind.ECMABoundary:
-                        return ECMABoundary;
-
                     case RegexNodeKind.Beginning:
-                        return Beginning;
-
                     case RegexNodeKind.Start:
-                        return Start;
-
                     case RegexNodeKind.EndZ:
-                        return EndZ;
-
                     case RegexNodeKind.End:
-                        return End;
+                    case RegexNodeKind.Boundary:
+                    case RegexNodeKind.ECMABoundary:
+                        // Return any anchor found.
+                        return node.Kind;
+
+                    case RegexNodeKind.Atomic:
+                    case RegexNodeKind.Capture:
+                        // For groups, continue exploring the sole child.
+                        node = node.Child(0);
+                        continue;
 
                     case RegexNodeKind.Concatenate:
-                        if (curNode.ChildCount() > 0)
+                        // For concatenations, we expect primarily to explore its last child, but we can also skip over
```

Comment:
NIT: this comment is only referring to the case where `leading == false`, perhaps we should modify it to say that if leading then we'll go forwards, and else go backwards.

Reply:
Oops, thanks.  I'd initially written a separate leading vs trailing implementation and then merged them and neglected to remove the comment.